mars/dataframe/indexing/set_index.py

```diff
@@ -0,0 +1,125 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+import six
+
+from ...serialize import AnyField, BoolField
+from ... import opcodes as OperandDef
+from ..operands import DataFrameOperand, DataFrameOperandMixin, ObjectType
+from ..utils import build_empty_df, parse_index
+
+
+class DataFrameSetIndex(DataFrameOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.DATAFRAME_SET_INDEX
+
+    _keys = AnyField('keys')
+    _drop = BoolField('drop')
+    _append = BoolField('append')
+    _verify_integrity = BoolField('verify_integrity')
+
+    def __init__(self, keys=None, drop=True, append=False, verify_integrity=False,
+                 object_type=None, **kw):
+        super(DataFrameSetIndex, self).__init__(_keys=keys, _drop=drop, _append=append,
+                                                _verify_integrity=verify_integrity, _object_type=object_type, **kw)
+
+    @property
+    def keys(self):
+        return self._keys
+
+    @property
+    def drop(self):
+        return self._drop
+
+    @property
+    def append(self):
+        return self._append
+
+    @property
+    def verify_integrity(self):
+        return self._verify_integrity
+
+    def __call__(self, df):
+        new_df = build_empty_df(df.dtypes).set_index(keys=self.keys, drop=self.drop, append=self.append,
+                                                     verify_integrity=self.verify_integrity)
+        return self.new_dataframe([df], shape=new_df.shape, dtypes=new_df.dtypes,
+                                  index_value=parse_index(new_df.index),
+                                  columns_value=parse_index(new_df.columns, store_data=True))
+
+    @classmethod
+    def tile(cls, op):
+        in_df = op.inputs[0]
+        out_df = op.outputs[0]
+
+        if not isinstance(op.keys, six.string_types):
+            raise NotImplementedError('DataFrame.set_index only support label')
+        if op.verify_integrity:
+            raise NotImplementedError('DataFrame.set_index not support verify_integrity yet')
+
+        out_chunks = []
+
+        try:
+            column_index = in_df.columns.to_pandas().get_loc(op.keys)
+        except KeyError:
+            raise NotImplementedError('The new index label must be a column of the original dataframe')
+
+        acc = 0
+        for idx, k in enumerate(in_df.nsplits[1]):
+            acc += k
+            if acc > column_index:
+                chunk_index = idx
+                break
+        else:
+            chunk_index = -1
```

Comment:
how about `np.searchsorted(np.cumsum(t), column_index)`?

Reply:
And it's better to provide a util as it can be used in `getitem`.

Reply:
Fixed, and thanks for the hint!